localstack/services/awslambda/invocation/docker_runtime_executor.py

```diff
@@ -218,12 +222,26 @@ def _build_executor_endpoint(self, service_endpoint: ServiceEndpoint) -> Executo
         )
         return executor_endpoint
 
+    def _generate_container_name(self):
+        """
+        Format <main-container-name>-lambda-<function-name>-<executor-id>
+        TODO: make the format configurable
+        """
+        container_name = "-".join(
+            [
+                get_main_container_name() or "localstack",
+                "lambda",
+                self.function_version.id.function_name.lower(),
+            ]
+        ).replace("_", "-")
```

Comment:
Do we want to standardize the container names (e.g., lowercase and `-` only) or make it easier to match for external lookups?

* `_` and capital letters would be valid for container names.
* `_` cannot be used in URLs and might lead to limitations with the DNS re-work (see Simon's post in #sig-localstack)

## Docker container names

* Validation regex: `[a-zA-Z0-9][a-zA-Z0-9_.-]` ([Docker source](https://github.com/moby/moby/blob/v23.0.1/daemon/names/names.go))
* Length constraint: There seem to be no explicit length constraints (also see [this SO discussion](https://stackoverflow.com/questions/42642561/docker-restrictions-regarding-naming-container)). There might be implicit limits (e.g., 63 chars in GitLab runners due to hitting max DNS hostname length mentioned [here](https://gitlab.com/gitlab-org/gitlab-runner/-/issues/27763))


## Input restrictions

* A Docker container name without length restriction should by definition be a valid input
* All AWS Lambda functions `[a-zA-Z0-9-_]+` should be valid container names (AWS [FunctionName](https://docs.aws.amazon.com/lambda/latest/dg/API_CreateFunction.html#SSS-CreateFunction-request-FunctionName) docs)

Reply:
Replacing `_` with `-` was also just partly because we do this in the k8s executor and partly for the DNS reason you mentioned. 

As you said, ideally we'd have names that can be used as valid DNS names, although the use case for this is pretty much non-existent for lambda at least and 63 Character is unfortunately also extremely limiting in practice, so we can't encode a lot of information this way. :/